# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of each sentence from `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    
    
    source = source_text.split("\n")
    target = target_text.split("\n")
    target = [s + " <EOS>" for s in target]
    
    source_id_text = list(map(lambda x: [source_vocab_to_int[word] for word in x.split()], source))
    target_id_text = list(map(lambda x: [target_vocab_to_int[word] for word in x.split()], target))
    
    return source_id_text, target_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf

# Check TensorFlow Version
assert LooseVersion(tf.__version__) in [LooseVersion('1.0.0'), LooseVersion('1.0.1')], 'This project requires TensorFlow version 1.0  You are using {}'.format(tf.__version__)
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.0.0
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoding_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.

Return the placeholders in the following the tuple (Input, Targets, Learing Rate, Keep Probability)

In [7]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, and learning rate.
    :return: Tuple (input, targets, learning rate, keep probability)
    """
    inputs = tf.placeholder(tf.int32, shape = [None, None], name="input")
    targets = tf.placeholder(tf.int32, shape = [None, None], name="targets")
    learning_rate = tf.placeholder(tf.float32, shape = [], name="learning_rate")
    keep_prob = tf.placeholder(tf.float32, shape = [], name="keep_prob")
        
    
    return inputs, targets, learning_rate, keep_prob

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoding Input
Implement `process_decoding_input` using TensorFlow to remove the last word id from each batch in `target_data` and concat the GO ID to the beginning of each batch.

In [8]:
def process_decoding_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for decoding
    :param target_data: Target Placeholder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    
    end = tf.strided_slice(target_data, [0,0], [batch_size,-1],[1,1])
    decoded_input = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int["<GO>"]), end],1)
    
    return decoded_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_decoding_input(process_decoding_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer using [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn).

In [9]:
def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :return: RNN state
    """
    LSTM = tf.contrib.rnn.BasicLSTMCell(rnn_size)
    RNN_cell = tf.contrib.rnn.MultiRNNCell(num_layers*[LSTM])
    _,RNN_state = tf.nn.dynamic_rnn(RNN_cell,rnn_inputs,dtype=tf.float32 )
    
    return RNN_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create training logits using [`tf.contrib.seq2seq.simple_decoder_fn_train()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/simple_decoder_fn_train) and [`tf.contrib.seq2seq.dynamic_rnn_decoder()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/dynamic_rnn_decoder).  Apply the `output_fn` to the [`tf.contrib.seq2seq.dynamic_rnn_decoder()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/dynamic_rnn_decoder) outputs.

In [10]:
def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, sequence_length, decoding_scope,
                         output_fn, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param sequence_length: Sequence Length
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_fn: Function to apply the output layer
    :param keep_prob: Dropout keep probability---------------[how to put in?]
    :return: Train Logits
    """
    
    train = tf.contrib.seq2seq.simple_decoder_fn_train(encoder_state)
    
    t_pred, _, _ = tf.contrib.seq2seq.dynamic_rnn_decoder(dec_cell , decoder_fn = train, 
                                                          inputs = dec_embed_input,
                                                          sequence_length = sequence_length,
                                                          scope = decoding_scope)
    
    train_logits = output_fn(t_pred)
    
    return train_logits


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference logits using [`tf.contrib.seq2seq.simple_decoder_fn_inference()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/simple_decoder_fn_inference) and [`tf.contrib.seq2seq.dynamic_rnn_decoder()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/dynamic_rnn_decoder). 

In [11]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id,
                         maximum_length, vocab_size, decoding_scope, output_fn, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param maximum_length: The maximum allowed time steps to decode
    :param vocab_size: Size of vocabulary
    :param decoding_scope: TensorFlow Variable Scope for decoding
    :param output_fn: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: Inference Logits
    """
    
    decoder_fn = tf.contrib.seq2seq.simple_decoder_fn_inference(output_fn, encoder_state,
                                                                dec_embeddings, start_of_sequence_id,
                                                                end_of_sequence_id, maximum_length,
                                                                vocab_size)
    
    inference_logits, _, _ = tf.contrib.seq2seq.dynamic_rnn_decoder(dec_cell,decoder_fn,scope=decoding_scope)
    
    return inference_logits


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

- Create RNN cell for decoding using `rnn_size` and `num_layers`.
- Create the output fuction using [`lambda`](https://docs.python.org/3/tutorial/controlflow.html#lambda-expressions) to transform it's input, logits, to class logits.
- Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, sequence_length, decoding_scope, output_fn, keep_prob)` function to get the training logits.
- Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, maximum_length, vocab_size, decoding_scope, output_fn, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [12]:
def decoding_layer(dec_embed_input, dec_embeddings, encoder_state, vocab_size, sequence_length, rnn_size,
                   num_layers, target_vocab_to_int, keep_prob):
    """
    Create decoding layer
    :param dec_embed_input: Decoder embedded input
    :param dec_embeddings: Decoder embeddings
    :param encoder_state: The encoded state
    :param vocab_size: Size of vocabulary
    :param sequence_length: Sequence Length
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param keep_prob: Dropout keep probability
    :return: Tuple of (Training Logits, Inference Logits)
    """
    
    with tf.variable_scope("decoder") as var_scope:
        
        output_fn = lambda x: tf.contrib.layers.fully_connected(x,vocab_size,None,scope=var_scope)
        
        LSTM = tf.contrib.rnn.BasicLSTMCell(rnn_size)
        LSTM_drop = tf.contrib.rnn.DropoutWrapper(LSTM,input_keep_prob=keep_prob)
        multi_LSTM_drop = tf.contrib.rnn.MultiRNNCell([LSTM_drop] * num_layers)
    
        Training_Logits = decoding_layer_train(encoder_state, multi_LSTM_drop, dec_embed_input,
                                              sequence_length, var_scope, output_fn, keep_prob)
        
        start = target_vocab_to_int["<GO>"]
        end = target_vocab_to_int["<EOS>"]
        
        var_scope.reuse_variables()
        
        Inference_Logits = decoding_layer_infer(encoder_state, multi_LSTM_drop, dec_embeddings,
                                                start, end, sequence_length, vocab_size, 
                                                var_scope, output_fn, keep_prob)
    
    return Training_Logits, Inference_Logits


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Apply embedding to the input data for the encoder.
- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob)`.
- Process target data using your `process_decoding_input(target_data, target_vocab_to_int, batch_size)` function.
- Apply embedding to the target data for the decoder.
- Decode the encoded input using your `decoding_layer(dec_embed_input, dec_embeddings, encoder_state, vocab_size, sequence_length, rnn_size, num_layers, target_vocab_to_int, keep_prob)`.

In [13]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size, sequence_length, source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size, rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param sequence_length: Sequence Length
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training Logits, Inference Logits)
    """
    
    inputs = tf.contrib.layers.embed_sequence(input_data, source_vocab_size, batch_size)
    encoder_state = encoding_layer(inputs,rnn_size, num_layers, keep_prob)
    dec_inputs = process_decoding_input(target_data, target_vocab_to_int, batch_size)
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size,dec_embedding_size]))
    dec_emb_inputs = tf.nn.embedding_lookup(dec_embeddings,dec_inputs)
    
    Training_Logits, Inference_Logits = decoding_layer(dec_emb_inputs, dec_embeddings, encoder_state,
                                                       target_vocab_size, sequence_length, rnn_size,
                                                       num_layers, target_vocab_to_int, keep_prob)
    
    
    return Training_Logits, Inference_Logits


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability

In [17]:
# Number of Epochs
epochs = 50
# Batch Size
batch_size = 4096
# RNN Size
rnn_size = 256
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 256
decoding_embedding_size = 256
# Learning Rate
learning_rate = 0.01
# Dropout Keep Probability
keep_probability = 0.7

### Build the Graph
Build the graph using the neural network you implemented.

In [18]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = '/home/iraptor/Documents/UdacityDeepLearning/3/deep-learning-master/language-translation'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_source_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob = model_inputs()
    sequence_length = tf.placeholder_with_default(max_source_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)
    
    train_logits, inference_logits = seq2seq_model(
        tf.reverse(input_data, [-1]), targets, keep_prob, batch_size, sequence_length, len(source_vocab_to_int), len(target_vocab_to_int),
        encoding_embedding_size, decoding_embedding_size, rnn_size, num_layers, target_vocab_to_int)

    tf.identity(inference_logits, 'logits')
    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            train_logits,
            targets,
            tf.ones([input_shape[0], sequence_length]))

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)

### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forums to see if anyone is having the same problem.

In [19]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import time

def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1]), (0,0)],
            'constant')

    return np.mean(np.equal(target, np.argmax(logits, 2)))

train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]

valid_source = helper.pad_sentence_batch(source_int_text[:batch_size])
valid_target = helper.pad_sentence_batch(target_int_text[:batch_size])

with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch) in enumerate(
                helper.batch_data(train_source, train_target, batch_size)):
            start_time = time.time()
            
            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 sequence_length: target_batch.shape[1],
                 keep_prob: keep_probability})
            
            batch_train_logits = sess.run(
                inference_logits,
                {input_data: source_batch, keep_prob: 1.0})
            batch_valid_logits = sess.run(
                inference_logits,
                {input_data: valid_source, keep_prob: 1.0})
                
            train_acc = get_accuracy(target_batch, batch_train_logits)
            valid_acc = get_accuracy(np.array(valid_target), batch_valid_logits)
            end_time = time.time()
            print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.3f}, Validation Accuracy: {:>6.3f}, Loss: {:>6.3f}'
                  .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    0/33 - Train Accuracy:  0.307, Validation Accuracy:  0.306, Loss:  5.896
Epoch   0 Batch    1/33 - Train Accuracy:  0.339, Validation Accuracy:  0.341, Loss:  4.125
Epoch   0 Batch    2/33 - Train Accuracy:  0.317, Validation Accuracy:  0.317, Loss:  4.100
Epoch   0 Batch    3/33 - Train Accuracy:  0.310, Validation Accuracy:  0.339, Loss:  3.608
Epoch   0 Batch    4/33 - Train Accuracy:  0.346, Validation Accuracy:  0.344, Loss:  3.367
Epoch   0 Batch    5/33 - Train Accuracy:  0.348, Validation Accuracy:  0.345, Loss:  3.247
Epoch   0 Batch    6/33 - Train Accuracy:  0.349, Validation Accuracy:  0.348, Loss:  3.205
Epoch   0 Batch    7/33 - Train Accuracy:  0.350, Validation Accuracy:  0.350, Loss:  3.134
Epoch   0 Batch    8/33 - Train Accuracy:  0.353, Validation Accuracy:  0.352, Loss:  3.039
Epoch   0 Batch    9/33 - Train Accuracy:  0.325, Validation Accuracy:  0.354, Loss:  3.048
Epoch   0 Batch   10/33 - Train Accuracy:  0.394, Validation Accuracy:  0.393, L

Epoch   2 Batch   26/33 - Train Accuracy:  0.635, Validation Accuracy:  0.627, Loss:  0.716
Epoch   2 Batch   27/33 - Train Accuracy:  0.667, Validation Accuracy:  0.630, Loss:  0.649
Epoch   2 Batch   28/33 - Train Accuracy:  0.620, Validation Accuracy:  0.633, Loss:  0.732
Epoch   2 Batch   29/33 - Train Accuracy:  0.654, Validation Accuracy:  0.636, Loss:  0.661
Epoch   2 Batch   30/33 - Train Accuracy:  0.644, Validation Accuracy:  0.640, Loss:  0.678
Epoch   2 Batch   31/33 - Train Accuracy:  0.637, Validation Accuracy:  0.640, Loss:  0.676
Epoch   3 Batch    0/33 - Train Accuracy:  0.645, Validation Accuracy:  0.642, Loss:  0.661
Epoch   3 Batch    1/33 - Train Accuracy:  0.654, Validation Accuracy:  0.645, Loss:  0.646
Epoch   3 Batch    2/33 - Train Accuracy:  0.650, Validation Accuracy:  0.651, Loss:  0.642
Epoch   3 Batch    3/33 - Train Accuracy:  0.643, Validation Accuracy:  0.653, Loss:  0.656
Epoch   3 Batch    4/33 - Train Accuracy:  0.659, Validation Accuracy:  0.658, L

Epoch   5 Batch   20/33 - Train Accuracy:  0.826, Validation Accuracy:  0.829, Loss:  0.260
Epoch   5 Batch   21/33 - Train Accuracy:  0.834, Validation Accuracy:  0.836, Loss:  0.249
Epoch   5 Batch   22/33 - Train Accuracy:  0.855, Validation Accuracy:  0.836, Loss:  0.214
Epoch   5 Batch   23/33 - Train Accuracy:  0.826, Validation Accuracy:  0.838, Loss:  0.244
Epoch   5 Batch   24/33 - Train Accuracy:  0.832, Validation Accuracy:  0.840, Loss:  0.241
Epoch   5 Batch   25/33 - Train Accuracy:  0.846, Validation Accuracy:  0.846, Loss:  0.228
Epoch   5 Batch   26/33 - Train Accuracy:  0.846, Validation Accuracy:  0.840, Loss:  0.221
Epoch   5 Batch   27/33 - Train Accuracy:  0.857, Validation Accuracy:  0.843, Loss:  0.201
Epoch   5 Batch   28/33 - Train Accuracy:  0.847, Validation Accuracy:  0.851, Loss:  0.223
Epoch   5 Batch   29/33 - Train Accuracy:  0.860, Validation Accuracy:  0.857, Loss:  0.201
Epoch   5 Batch   30/33 - Train Accuracy:  0.854, Validation Accuracy:  0.851, L

Epoch   8 Batch   14/33 - Train Accuracy:  0.930, Validation Accuracy:  0.936, Loss:  0.065
Epoch   8 Batch   15/33 - Train Accuracy:  0.930, Validation Accuracy:  0.938, Loss:  0.065
Epoch   8 Batch   16/33 - Train Accuracy:  0.932, Validation Accuracy:  0.935, Loss:  0.064
Epoch   8 Batch   17/33 - Train Accuracy:  0.939, Validation Accuracy:  0.936, Loss:  0.062
Epoch   8 Batch   18/33 - Train Accuracy:  0.938, Validation Accuracy:  0.940, Loss:  0.061
Epoch   8 Batch   19/33 - Train Accuracy:  0.939, Validation Accuracy:  0.941, Loss:  0.059
Epoch   8 Batch   20/33 - Train Accuracy:  0.940, Validation Accuracy:  0.941, Loss:  0.062
Epoch   8 Batch   21/33 - Train Accuracy:  0.934, Validation Accuracy:  0.944, Loss:  0.060
Epoch   8 Batch   22/33 - Train Accuracy:  0.949, Validation Accuracy:  0.942, Loss:  0.050
Epoch   8 Batch   23/33 - Train Accuracy:  0.934, Validation Accuracy:  0.940, Loss:  0.059
Epoch   8 Batch   24/33 - Train Accuracy:  0.938, Validation Accuracy:  0.941, L

Epoch  11 Batch    8/33 - Train Accuracy:  0.959, Validation Accuracy:  0.959, Loss:  0.032
Epoch  11 Batch    9/33 - Train Accuracy:  0.954, Validation Accuracy:  0.959, Loss:  0.031
Epoch  11 Batch   10/33 - Train Accuracy:  0.963, Validation Accuracy:  0.960, Loss:  0.029
Epoch  11 Batch   11/33 - Train Accuracy:  0.959, Validation Accuracy:  0.961, Loss:  0.032
Epoch  11 Batch   12/33 - Train Accuracy:  0.962, Validation Accuracy:  0.961, Loss:  0.030
Epoch  11 Batch   13/33 - Train Accuracy:  0.961, Validation Accuracy:  0.960, Loss:  0.028
Epoch  11 Batch   14/33 - Train Accuracy:  0.959, Validation Accuracy:  0.960, Loss:  0.030
Epoch  11 Batch   15/33 - Train Accuracy:  0.960, Validation Accuracy:  0.963, Loss:  0.028
Epoch  11 Batch   16/33 - Train Accuracy:  0.959, Validation Accuracy:  0.964, Loss:  0.029
Epoch  11 Batch   17/33 - Train Accuracy:  0.965, Validation Accuracy:  0.963, Loss:  0.028
Epoch  11 Batch   18/33 - Train Accuracy:  0.963, Validation Accuracy:  0.964, L

Epoch  14 Batch    2/33 - Train Accuracy:  0.970, Validation Accuracy:  0.968, Loss:  0.019
Epoch  14 Batch    3/33 - Train Accuracy:  0.972, Validation Accuracy:  0.969, Loss:  0.019
Epoch  14 Batch    4/33 - Train Accuracy:  0.971, Validation Accuracy:  0.968, Loss:  0.018
Epoch  14 Batch    5/33 - Train Accuracy:  0.974, Validation Accuracy:  0.969, Loss:  0.017
Epoch  14 Batch    6/33 - Train Accuracy:  0.970, Validation Accuracy:  0.968, Loss:  0.019
Epoch  14 Batch    7/33 - Train Accuracy:  0.970, Validation Accuracy:  0.968, Loss:  0.020
Epoch  14 Batch    8/33 - Train Accuracy:  0.971, Validation Accuracy:  0.968, Loss:  0.020
Epoch  14 Batch    9/33 - Train Accuracy:  0.968, Validation Accuracy:  0.971, Loss:  0.020
Epoch  14 Batch   10/33 - Train Accuracy:  0.974, Validation Accuracy:  0.969, Loss:  0.018
Epoch  14 Batch   11/33 - Train Accuracy:  0.969, Validation Accuracy:  0.970, Loss:  0.019
Epoch  14 Batch   12/33 - Train Accuracy:  0.971, Validation Accuracy:  0.970, L

Epoch  16 Batch   28/33 - Train Accuracy:  0.978, Validation Accuracy:  0.973, Loss:  0.013
Epoch  16 Batch   29/33 - Train Accuracy:  0.977, Validation Accuracy:  0.975, Loss:  0.014
Epoch  16 Batch   30/33 - Train Accuracy:  0.980, Validation Accuracy:  0.974, Loss:  0.013
Epoch  16 Batch   31/33 - Train Accuracy:  0.979, Validation Accuracy:  0.976, Loss:  0.014
Epoch  17 Batch    0/33 - Train Accuracy:  0.978, Validation Accuracy:  0.973, Loss:  0.013
Epoch  17 Batch    1/33 - Train Accuracy:  0.979, Validation Accuracy:  0.974, Loss:  0.013
Epoch  17 Batch    2/33 - Train Accuracy:  0.975, Validation Accuracy:  0.973, Loss:  0.013
Epoch  17 Batch    3/33 - Train Accuracy:  0.979, Validation Accuracy:  0.974, Loss:  0.014
Epoch  17 Batch    4/33 - Train Accuracy:  0.979, Validation Accuracy:  0.975, Loss:  0.013
Epoch  17 Batch    5/33 - Train Accuracy:  0.980, Validation Accuracy:  0.973, Loss:  0.012
Epoch  17 Batch    6/33 - Train Accuracy:  0.977, Validation Accuracy:  0.974, L

Epoch  19 Batch   22/33 - Train Accuracy:  0.984, Validation Accuracy:  0.978, Loss:  0.010
Epoch  19 Batch   23/33 - Train Accuracy:  0.985, Validation Accuracy:  0.977, Loss:  0.010
Epoch  19 Batch   24/33 - Train Accuracy:  0.980, Validation Accuracy:  0.978, Loss:  0.011
Epoch  19 Batch   25/33 - Train Accuracy:  0.983, Validation Accuracy:  0.978, Loss:  0.011
Epoch  19 Batch   26/33 - Train Accuracy:  0.978, Validation Accuracy:  0.977, Loss:  0.011
Epoch  19 Batch   27/33 - Train Accuracy:  0.983, Validation Accuracy:  0.976, Loss:  0.010
Epoch  19 Batch   28/33 - Train Accuracy:  0.982, Validation Accuracy:  0.977, Loss:  0.011
Epoch  19 Batch   29/33 - Train Accuracy:  0.982, Validation Accuracy:  0.977, Loss:  0.011
Epoch  19 Batch   30/33 - Train Accuracy:  0.983, Validation Accuracy:  0.976, Loss:  0.009
Epoch  19 Batch   31/33 - Train Accuracy:  0.981, Validation Accuracy:  0.978, Loss:  0.011
Epoch  20 Batch    0/33 - Train Accuracy:  0.985, Validation Accuracy:  0.978, L

Epoch  22 Batch   16/33 - Train Accuracy:  0.983, Validation Accuracy:  0.981, Loss:  0.009
Epoch  22 Batch   17/33 - Train Accuracy:  0.987, Validation Accuracy:  0.979, Loss:  0.008
Epoch  22 Batch   18/33 - Train Accuracy:  0.985, Validation Accuracy:  0.980, Loss:  0.009
Epoch  22 Batch   19/33 - Train Accuracy:  0.986, Validation Accuracy:  0.980, Loss:  0.008
Epoch  22 Batch   20/33 - Train Accuracy:  0.984, Validation Accuracy:  0.979, Loss:  0.009
Epoch  22 Batch   21/33 - Train Accuracy:  0.984, Validation Accuracy:  0.979, Loss:  0.010
Epoch  22 Batch   22/33 - Train Accuracy:  0.986, Validation Accuracy:  0.979, Loss:  0.007
Epoch  22 Batch   23/33 - Train Accuracy:  0.985, Validation Accuracy:  0.979, Loss:  0.008
Epoch  22 Batch   24/33 - Train Accuracy:  0.985, Validation Accuracy:  0.979, Loss:  0.008
Epoch  22 Batch   25/33 - Train Accuracy:  0.986, Validation Accuracy:  0.980, Loss:  0.008
Epoch  22 Batch   26/33 - Train Accuracy:  0.983, Validation Accuracy:  0.979, L

Epoch  25 Batch   10/33 - Train Accuracy:  0.984, Validation Accuracy:  0.978, Loss:  0.007
Epoch  25 Batch   11/33 - Train Accuracy:  0.984, Validation Accuracy:  0.980, Loss:  0.008
Epoch  25 Batch   12/33 - Train Accuracy:  0.983, Validation Accuracy:  0.977, Loss:  0.007
Epoch  25 Batch   13/33 - Train Accuracy:  0.984, Validation Accuracy:  0.978, Loss:  0.008
Epoch  25 Batch   14/33 - Train Accuracy:  0.980, Validation Accuracy:  0.976, Loss:  0.008
Epoch  25 Batch   15/33 - Train Accuracy:  0.985, Validation Accuracy:  0.982, Loss:  0.009
Epoch  25 Batch   16/33 - Train Accuracy:  0.977, Validation Accuracy:  0.975, Loss:  0.008
Epoch  25 Batch   17/33 - Train Accuracy:  0.987, Validation Accuracy:  0.981, Loss:  0.008
Epoch  25 Batch   18/33 - Train Accuracy:  0.981, Validation Accuracy:  0.977, Loss:  0.008
Epoch  25 Batch   19/33 - Train Accuracy:  0.986, Validation Accuracy:  0.980, Loss:  0.008
Epoch  25 Batch   20/33 - Train Accuracy:  0.982, Validation Accuracy:  0.977, L

Epoch  28 Batch    4/33 - Train Accuracy:  0.988, Validation Accuracy:  0.981, Loss:  0.006
Epoch  28 Batch    5/33 - Train Accuracy:  0.987, Validation Accuracy:  0.979, Loss:  0.006
Epoch  28 Batch    6/33 - Train Accuracy:  0.987, Validation Accuracy:  0.980, Loss:  0.007
Epoch  28 Batch    7/33 - Train Accuracy:  0.984, Validation Accuracy:  0.979, Loss:  0.007
Epoch  28 Batch    8/33 - Train Accuracy:  0.987, Validation Accuracy:  0.982, Loss:  0.007
Epoch  28 Batch    9/33 - Train Accuracy:  0.987, Validation Accuracy:  0.982, Loss:  0.006
Epoch  28 Batch   10/33 - Train Accuracy:  0.988, Validation Accuracy:  0.981, Loss:  0.006
Epoch  28 Batch   11/33 - Train Accuracy:  0.984, Validation Accuracy:  0.982, Loss:  0.007
Epoch  28 Batch   12/33 - Train Accuracy:  0.988, Validation Accuracy:  0.982, Loss:  0.006
Epoch  28 Batch   13/33 - Train Accuracy:  0.987, Validation Accuracy:  0.981, Loss:  0.006
Epoch  28 Batch   14/33 - Train Accuracy:  0.986, Validation Accuracy:  0.982, L

Epoch  30 Batch   30/33 - Train Accuracy:  0.989, Validation Accuracy:  0.981, Loss:  0.005
Epoch  30 Batch   31/33 - Train Accuracy:  0.987, Validation Accuracy:  0.982, Loss:  0.006
Epoch  31 Batch    0/33 - Train Accuracy:  0.988, Validation Accuracy:  0.981, Loss:  0.006
Epoch  31 Batch    1/33 - Train Accuracy:  0.987, Validation Accuracy:  0.981, Loss:  0.005
Epoch  31 Batch    2/33 - Train Accuracy:  0.988, Validation Accuracy:  0.981, Loss:  0.006
Epoch  31 Batch    3/33 - Train Accuracy:  0.987, Validation Accuracy:  0.981, Loss:  0.005
Epoch  31 Batch    4/33 - Train Accuracy:  0.988, Validation Accuracy:  0.981, Loss:  0.006
Epoch  31 Batch    5/33 - Train Accuracy:  0.990, Validation Accuracy:  0.981, Loss:  0.005
Epoch  31 Batch    6/33 - Train Accuracy:  0.988, Validation Accuracy:  0.982, Loss:  0.006
Epoch  31 Batch    7/33 - Train Accuracy:  0.987, Validation Accuracy:  0.983, Loss:  0.006
Epoch  31 Batch    8/33 - Train Accuracy:  0.988, Validation Accuracy:  0.983, L

Epoch  33 Batch   24/33 - Train Accuracy:  0.989, Validation Accuracy:  0.982, Loss:  0.005
Epoch  33 Batch   25/33 - Train Accuracy:  0.988, Validation Accuracy:  0.983, Loss:  0.005
Epoch  33 Batch   26/33 - Train Accuracy:  0.987, Validation Accuracy:  0.982, Loss:  0.006
Epoch  33 Batch   27/33 - Train Accuracy:  0.989, Validation Accuracy:  0.981, Loss:  0.005
Epoch  33 Batch   28/33 - Train Accuracy:  0.988, Validation Accuracy:  0.982, Loss:  0.005
Epoch  33 Batch   29/33 - Train Accuracy:  0.987, Validation Accuracy:  0.983, Loss:  0.005
Epoch  33 Batch   30/33 - Train Accuracy:  0.988, Validation Accuracy:  0.982, Loss:  0.005
Epoch  33 Batch   31/33 - Train Accuracy:  0.989, Validation Accuracy:  0.983, Loss:  0.006
Epoch  34 Batch    0/33 - Train Accuracy:  0.988, Validation Accuracy:  0.981, Loss:  0.005
Epoch  34 Batch    1/33 - Train Accuracy:  0.990, Validation Accuracy:  0.982, Loss:  0.005
Epoch  34 Batch    2/33 - Train Accuracy:  0.989, Validation Accuracy:  0.982, L

Epoch  36 Batch   18/33 - Train Accuracy:  0.990, Validation Accuracy:  0.981, Loss:  0.005
Epoch  36 Batch   19/33 - Train Accuracy:  0.991, Validation Accuracy:  0.983, Loss:  0.004
Epoch  36 Batch   20/33 - Train Accuracy:  0.989, Validation Accuracy:  0.983, Loss:  0.004
Epoch  36 Batch   21/33 - Train Accuracy:  0.989, Validation Accuracy:  0.982, Loss:  0.005
Epoch  36 Batch   22/33 - Train Accuracy:  0.991, Validation Accuracy:  0.984, Loss:  0.004
Epoch  36 Batch   23/33 - Train Accuracy:  0.989, Validation Accuracy:  0.981, Loss:  0.004
Epoch  36 Batch   24/33 - Train Accuracy:  0.991, Validation Accuracy:  0.981, Loss:  0.005
Epoch  36 Batch   25/33 - Train Accuracy:  0.989, Validation Accuracy:  0.983, Loss:  0.005
Epoch  36 Batch   26/33 - Train Accuracy:  0.988, Validation Accuracy:  0.981, Loss:  0.005
Epoch  36 Batch   27/33 - Train Accuracy:  0.991, Validation Accuracy:  0.983, Loss:  0.005
Epoch  36 Batch   28/33 - Train Accuracy:  0.989, Validation Accuracy:  0.981, L

Epoch  39 Batch   12/33 - Train Accuracy:  0.990, Validation Accuracy:  0.982, Loss:  0.004
Epoch  39 Batch   13/33 - Train Accuracy:  0.990, Validation Accuracy:  0.982, Loss:  0.004
Epoch  39 Batch   14/33 - Train Accuracy:  0.988, Validation Accuracy:  0.983, Loss:  0.005
Epoch  39 Batch   15/33 - Train Accuracy:  0.988, Validation Accuracy:  0.983, Loss:  0.005
Epoch  39 Batch   16/33 - Train Accuracy:  0.990, Validation Accuracy:  0.984, Loss:  0.005
Epoch  39 Batch   17/33 - Train Accuracy:  0.991, Validation Accuracy:  0.984, Loss:  0.004
Epoch  39 Batch   18/33 - Train Accuracy:  0.989, Validation Accuracy:  0.982, Loss:  0.004
Epoch  39 Batch   19/33 - Train Accuracy:  0.991, Validation Accuracy:  0.981, Loss:  0.004
Epoch  39 Batch   20/33 - Train Accuracy:  0.990, Validation Accuracy:  0.985, Loss:  0.004
Epoch  39 Batch   21/33 - Train Accuracy:  0.991, Validation Accuracy:  0.984, Loss:  0.005
Epoch  39 Batch   22/33 - Train Accuracy:  0.990, Validation Accuracy:  0.984, L

Epoch  42 Batch    6/33 - Train Accuracy:  0.991, Validation Accuracy:  0.984, Loss:  0.004
Epoch  42 Batch    7/33 - Train Accuracy:  0.992, Validation Accuracy:  0.985, Loss:  0.004
Epoch  42 Batch    8/33 - Train Accuracy:  0.991, Validation Accuracy:  0.985, Loss:  0.004
Epoch  42 Batch    9/33 - Train Accuracy:  0.993, Validation Accuracy:  0.986, Loss:  0.004
Epoch  42 Batch   10/33 - Train Accuracy:  0.992, Validation Accuracy:  0.987, Loss:  0.004
Epoch  42 Batch   11/33 - Train Accuracy:  0.991, Validation Accuracy:  0.985, Loss:  0.004
Epoch  42 Batch   12/33 - Train Accuracy:  0.992, Validation Accuracy:  0.984, Loss:  0.004
Epoch  42 Batch   13/33 - Train Accuracy:  0.991, Validation Accuracy:  0.983, Loss:  0.004
Epoch  42 Batch   14/33 - Train Accuracy:  0.990, Validation Accuracy:  0.984, Loss:  0.004
Epoch  42 Batch   15/33 - Train Accuracy:  0.991, Validation Accuracy:  0.984, Loss:  0.004
Epoch  42 Batch   16/33 - Train Accuracy:  0.990, Validation Accuracy:  0.983, L

Epoch  45 Batch    0/33 - Train Accuracy:  0.991, Validation Accuracy:  0.984, Loss:  0.004
Epoch  45 Batch    1/33 - Train Accuracy:  0.992, Validation Accuracy:  0.984, Loss:  0.004
Epoch  45 Batch    2/33 - Train Accuracy:  0.992, Validation Accuracy:  0.983, Loss:  0.004
Epoch  45 Batch    3/33 - Train Accuracy:  0.992, Validation Accuracy:  0.982, Loss:  0.003
Epoch  45 Batch    4/33 - Train Accuracy:  0.993, Validation Accuracy:  0.983, Loss:  0.004
Epoch  45 Batch    5/33 - Train Accuracy:  0.992, Validation Accuracy:  0.983, Loss:  0.003
Epoch  45 Batch    6/33 - Train Accuracy:  0.989, Validation Accuracy:  0.982, Loss:  0.004
Epoch  45 Batch    7/33 - Train Accuracy:  0.991, Validation Accuracy:  0.984, Loss:  0.005
Epoch  45 Batch    8/33 - Train Accuracy:  0.991, Validation Accuracy:  0.982, Loss:  0.004
Epoch  45 Batch    9/33 - Train Accuracy:  0.992, Validation Accuracy:  0.982, Loss:  0.004
Epoch  45 Batch   10/33 - Train Accuracy:  0.992, Validation Accuracy:  0.985, L

Epoch  47 Batch   26/33 - Train Accuracy:  0.992, Validation Accuracy:  0.984, Loss:  0.003
Epoch  47 Batch   27/33 - Train Accuracy:  0.993, Validation Accuracy:  0.982, Loss:  0.004
Epoch  47 Batch   28/33 - Train Accuracy:  0.993, Validation Accuracy:  0.981, Loss:  0.003
Epoch  47 Batch   29/33 - Train Accuracy:  0.992, Validation Accuracy:  0.982, Loss:  0.003
Epoch  47 Batch   30/33 - Train Accuracy:  0.993, Validation Accuracy:  0.982, Loss:  0.003
Epoch  47 Batch   31/33 - Train Accuracy:  0.992, Validation Accuracy:  0.983, Loss:  0.004
Epoch  48 Batch    0/33 - Train Accuracy:  0.992, Validation Accuracy:  0.984, Loss:  0.004
Epoch  48 Batch    1/33 - Train Accuracy:  0.992, Validation Accuracy:  0.983, Loss:  0.004
Epoch  48 Batch    2/33 - Train Accuracy:  0.992, Validation Accuracy:  0.983, Loss:  0.003
Epoch  48 Batch    3/33 - Train Accuracy:  0.993, Validation Accuracy:  0.984, Loss:  0.003
Epoch  48 Batch    4/33 - Train Accuracy:  0.993, Validation Accuracy:  0.985, L

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [20]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [21]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
- Convert words not in the vocabulary, to the `<UNK>` word id.

In [28]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    sent = sentence.lower().split()
    seq = []
    
    for w in sent:
        if w in vocab_to_int:
            seq.append(vocab_to_int[w])
        else:
            seq.append(vocab_to_int['<UNK>'])
    
    return seq
    


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [29]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('logits:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence], keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in np.argmax(translate_logits, 1)]))
print('  French Words: {}'.format([target_int_to_vocab[i] for i in np.argmax(translate_logits, 1)]))

Input
  Word Ids:      [139, 192, 20, 134, 53, 182, 166]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [301, 91, 263, 269, 351, 336, 314, 338, 1]
  French Words: ['il', 'a', 'vu', 'un', 'vieux', 'camion', 'espagnol', '.', '<EOS>']


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  Additionally, the translations in this data set were made by Google translate, so the translations themselves aren't particularly good.  (We apologize to the French speakers out there!) Thankfully, for this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.